windows/src/desktop/kmshell/main/Keyman.System.RemoteUpdateCheck.pas

```diff
@@ -0,0 +1,386 @@
+  Name:             WebUpdateCheck
+  Copyright:        Copyright (C) SIL International.
+  Documentation:
+  Description:
+  Create Date:      5 Dec 2023
+
+  Modified Date:
+  Authors:          rcruickshank
+  Related Files:
+  Dependencies:
+
+  Bugs:
+  Todo:
+  Notes:
+  History:
+*)
+unit Keyman.System.RemoteUpdateCheck;  // I3306
+
+interface
+
+uses
+  System.Classes,
+  System.SysUtils,
+  KeymanPaths,
+  httpuploader,
+  Keyman.System.UpdateCheckResponse,
+  OnlineUpdateCheck;
+
+type
+  ERemoteUpdateCheck = class(Exception);
+
+  TRemoteUpdateCheckResult = (wucUnknown, wucSuccess, wucNoUpdates, wucFailure, wucOffline);
+
+  TRemoteUpdateCheckDownloadParams = record
+    TotalSize: Integer;
+    TotalDownloads: Integer;
+    StartPosition: Integer;
+  end;
+
+  TRemoteUpdateCheck = class
+  private
+    FForce: Boolean;
+    FRemoteResult: TRemoteUpdateCheckResult;
+
+    FErrorMessage: string;
+
+    FShowErrors: Boolean;
+    FDownload: TRemoteUpdateCheckDownloadParams;
+    FCheckOnly: Boolean;
+
+    function DownloadUpdates(Params: TUpdateCheckResponse) : Boolean;
+    procedure DoDownloadUpdates(SavePath: string; Params: TUpdateCheckResponse;  var Result: Boolean);
+    function DoRun: TRemoteUpdateCheckResult;
+  public
+
+    constructor Create(AForce : Boolean; ACheckOnly: Boolean = False);
+    destructor Destroy; override;
+    function Run: TRemoteUpdateCheckResult;
+    property ShowErrors: Boolean read FShowErrors write FShowErrors;
+  end;
+
+procedure LogMessage(LogMessage: string);
+
+implementation
+
+uses
+  System.WideStrUtils,
+  Winapi.Windows,
+  Winapi.WinINet,
+
+  GlobalProxySettings,
+  KLog,
+  keymanapi_TLB,
+  KeymanVersion,
+  Keyman.System.UpdateCheckStorage,
+  kmint,
+  ErrorControlledRegistry,
+  RegistryKeys,
+  Upload_Settings,
+
+  OnlineUpdateCheckMessages;
+
+{ TRemoteUpdateCheck }
+
+constructor TRemoteUpdateCheck.Create(AForce, ACheckOnly: Boolean);
+begin
+  inherited Create;
+
+  FShowErrors := True;
+  FRemoteResult := wucUnknown;
+
+  FForce := AForce;
+  FCheckOnly := ACheckOnly;
+
+  KL.Log('TRemoteUpdateCheck.Create');
+end;
+
+destructor TRemoteUpdateCheck.Destroy;
+begin
+  if (FErrorMessage <> '') and FShowErrors then
+    LogMessage(FErrorMessage);
+
+  KL.Log('TRemoteUpdateCheck.Destroy: FErrorMessage = '+FErrorMessage);
+  KL.Log('TRemoteUpdateCheck.Destroy: FRemoteResult = '+IntToStr(Ord(FRemoteResult)));
+
+  inherited Destroy;
+end;
+
+function TRemoteUpdateCheck.Run: TRemoteUpdateCheckResult;
+begin
+  Result := DoRun;
+
+  if Result in [ wucSuccess] then
+  begin
+    kmcom.Keyboards.Refresh;
+    kmcom.Keyboards.Apply;
+    kmcom.Packages.Refresh;
+  end;
+
+  FRemoteResult := Result;
+end;
+
+
+procedure TRemoteUpdateCheck.DoDownloadUpdates(SavePath: string; Params: TUpdateCheckResponse; var Result: Boolean);
+var
+  i, downloadCount: Integer;
+
+    function DownloadFile(const url, savepath: string): Boolean;
+    begin
+      with THttpUploader.Create(nil) do
+      try
```

Comment:
Note that we are trying to move away from `with` because it lacks scope clarity. Instead, use:

```suggestion
    var
      http: THttpUploader;
    begin
      http := THttpUploader.Create(nil);
      try
```

and then `http.Proxy.Server := ...` etc.

and don't forget that:

```
finally
  http.Free;
end;
```